### PR TITLE
added methods Table.with_relabeling and Table.with_column

### PR DESCRIPTION
**Pull request by ericz82**
#118

### CODE REVIEW

**Comment by SamLau95** [comment]
Cool! Couple things:

- Let's add documentation and tests for these methods
- Let's make these methods also be able to take in a sequence as well as an individual value. (Eg. `with_relabeling(['A', 'B', 'C'], ['a', 'b', 'c'])` should relabel all the columns specified.)

**Comment by ericz82** [comment]
How do I add tests?

**Comment by mDibyo** [comment]
you can add tests at [tests/test_tables.py](https://github.com/dsten/datascience/blob/master/tests/test_tables.py)

File: datascience/tables.py [hunk 1]
```diff
@@ -395,6 +395,13 @@ def append_column(self, label, values):
         self._columns[label] = values
         return self
 
+    def with_column(self, label, values):
+        """Returns a new table with new column included."""
```
Comment (stefanv) [inline datascience/tables.py]: This docstring is rather vague.

Comment (stefanv) [inline datascience/tables.py]: Also, what does it add above `append_column`?  If this is just to make a copy, we should rather implement ``table.copy()``.

File: datascience/tables.py [hunk 2]
```diff
@@ -406,6 +413,12 @@ def relabel(self, column_label, new_label):
             self._formats[new_label] = formatter
         return self
 
+    def with_relabeling(self, column_label, new_label):
+        """Returns a new table with column_label changed to new_label."""
```
Comment (stefanv) [inline datascience/tables.py]: This docstring should be expanded along the lines of the others.  It should describe that the parameters expected can be a single string, or a tuple of strings.  Perhaps it can also take a dict?  Alternatives to the name ``with_relabeling`` could be ``table.relabel``, ``table.rename`` or ``table.rename_columns``.

File: docs/_autosummary/datascience.tables.Table.as_html.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+datascience.tables.Table.as_html
```
Comment (stefanv) [inline docs/_autosummary/datascience.tables.Table.as_html.rst]: These files should not be included in the PR.

**Comment by ericz82** [comment]
Should I change Table.relabel to also take in a sequence of arguments?

**Comment by stefanv** [comment]
@ericz82 I think it already does (according to the example you gave).  It should simply be documented that way.

**Comment by ericz82** [comment]
@stefanv Hmm, I'm not sure what example you are referring to. I wrote a test for Table.relabel taking in a list for ``column_label`` and a list for ``new_label`` which does not seem to be passing. I believe Table.relabel should be changed accordingly.

**Comment by stefanv** [comment]
Ah!  But there you go.  You should just rework `relabel` to have the functionality you want, instead of adding a whole new method.

**Comment by ericz82** [comment]
Yeah, that's what I decided on doing. Everything should be good now.

File: datascience/tables.py [hunk 4]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
+        a      | 9     | 1      | 10       | hello
+        b      | 3     | 2      | 20       | hello
+        c      | 3     | 2      | 30       | hello
+        z      | 1     | 10     | 40       | hello
+        >>> table
+        letter | count | points
+        a      | 9     | 1
+        b      | 3     | 2
+        c      | 3     | 2
+        z      | 1     | 10
```
Comment (SamLau95) [inline datascience/tables.py]: Good that you showed that `table` is unchanged! But you wrote `table.append_column('new_col1', [10, 20, 30, 40])` instead of `table.with_column('new_col1', [10, 20, 30, 40])`. :)

File: docs/tables.rst [hunk 5]
```diff
@@ -74,7 +74,9 @@ Mutation (modifies table in place)
     Table.move_to_end
     Table.append
     Table.append_column
+    Table.with_column
     Table.relabel
+    Table.with_relabeling
```
Comment (SamLau95) [inline docs/tables.rst]: These belong in the Transformation section because they return a new table instead of mutating the original!

File: datascience/tables.py [hunk 6]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
         self._columns[label] = values
         return self
 
+    def with_column(self, label, values):
+        """Returns a separate new table with new column included.
```
Comment (SamLau95) [inline datascience/tables.py]: This can just say `Returns a new table with new column included.`

File: datascience/tables.py [hunk 7]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
-        assert column_label in self._columns
-        rewrite = lambda s: new_label if s == column_label else s
+        """Change the labels of columns specified by ``column_label`` to
+        labels in ``new_label``.
+
+        Args:
+            ``column_label`` (single str or list/array of str): The label(s) of
+                columns to be changed. Must be str.
+
+            ``new_label`` (single str or list/array of): The new label(s) of
```
Comment (SamLau95) [inline datascience/tables.py]: nit: `list/array of str`

File: datascience/tables.py [hunk 8]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
+                columns to be changed. Must be str.
+
+                Number of elements must match number of elements in
+                ``column_label``.
```
Comment (SamLau95) [inline datascience/tables.py]: If the lengths don't match up, let's throw a ValueError. See `Table.split` for an example.

File: datascience/tables.py [hunk 9]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
+
+        Returns:
+            None
```
Comment (SamLau95) [inline datascience/tables.py]: This actually returns the original table, so let's write that here.

File: docs/tables.rst (hunk 5 above)
Comment (ericz82) [inline docs/tables.rst]: @SamLau95 Should I also change the locations of these methods in tables.py?

**Comment by stefanv** [comment]
So, API question: is the new method added here then equivalent to:

```
t1= t0.copy()
t1.relabel(['a','b','c'], ['A','B','C'])
```

?  If so, is it worth expanding the API? (I have no idea how common such an operation is.)

File: datascience/tables.py [hunk 10]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
+        >>> table
+        red    | blue
+        1      | 12345
+        2      | 123
+        3      | 5123
+        """
+        if isinstance(column_label, str) and isinstance(new_label, str):
+            column_label, new_label = [column_label], [new_label]
+        d = dict(zip(column_label, new_label)) # dictioanry to map old labels to new ones
+        for col in d.keys():
```
Comment (SamLau95) [inline datascience/tables.py]: Let's call this dictionary `old_to_new` instead of `d` since it'll help someone who's reading this for the first time. Also, dictionaries iterate over their keys by default so you can more succinctly write `for col in old_to_new:`.

nit: You misspelled `dictioanry`!

File: datascience/tables.py [hunk 11]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
+            assert col in self._columns
```
Comment (SamLau95) [inline datascience/tables.py]: A ValueError is be better to raise here!

File: datascience/tables.py [hunk 12]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
+        rewrite = lambda s: d[s] if s in d.keys() else s
```
Comment (SamLau95) [inline datascience/tables.py]: nit: As with above, you can omit the `.keys()`

File: datascience/tables.py [hunk 13]
```diff
@@ -395,17 +395,156 @@ def append_column(self, label, values):
...
-            self._formats[new_label] = formatter
+        for label in self._formats:
+            if label in column_label:
+                formatter = self._formats.pop(label)
+                self._formats[d[label]] = formatter
         return self
 
+    def with_relabeling(self, column_label, new_label):
+        """Returns a new table with column_label changed to new_label.
+        ``column_label`` and ``new_label`` may be single values or lists
```
Comment (SamLau95) [inline datascience/tables.py]: Let's insert a newline above the line starting with column_label and new_label so that the summary on the main table doc page is short.

File: docs/tables.rst (hunk 5 above)
Comment (SamLau95) [inline docs/tables.rst]: Good catch! Please do that.

**Comment by SamLau95** [comment]
@stefanv I think it is worth expanding the API in this case because method chaining is much easier with these two methods. In addition, I believe we're going to deprecate the mutation methods (or at least not teach them) in favor of these new ones because of the issues we face with students running cells multiple times. See https://github.com/dsten/datascience/issues/114

**Comment by stefanv** [comment]
@SamLau95 Thanks, that makes sense.  I think there are a couple of sneaky places in which data mutation can still accidentally happen.  If this is a decided policy (never to mutate), then we can start hunting for those.

**Comment by SamLau95** [comment]
@stefanv I'm not sure if it's a decided policy but it definitely helps a lot with reasoning about tables, especially for students that are just starting to learn programming. Originally we were fine with mutation but I think the de facto policy now is just to replace mutation with a pure function alternative as much as we can.

**Comment by SamLau95** [comment]
Thanks for the hard work on this @ericz82 !
